README.md
LGeneral-Deluxe
===============

Повторная реализация стратегической игры Panzer General с открытым исходным кодом.


ChangeLog
- добавлено окно сообщения с чит-кодами (2014/07/06 AndO3131)
- исправить маленькую карту при падении с высоким разрешением экрана (2014/04/24 AndO3131)
- окно сообщения позволяет отображать и писать текст в игре (2014/01/26 AndO3131)
- Редизайн графического интерфейса - 3 информационных поля вверху, отображается дополнительная информация об игре
  (2014/01/15 http://openxcom.org/ AndO3131)
- движок способен загружать папки кампании PGF (2014/01/14 AndO3131)
- исправлена ​​потеря наземного транспортера при заходе в море (2014/01/02 AndO3131)
- игрок может выбрать папку для загрузки игровых данных (2013/12/19 AndO3131)
- экран загрузки / сохранения, позволяющий несколько сохранений (2013/12/18 AndO3131)
- Добавлен установщик NSIS (2013/12/11 mas1904)
- добавлен выбор для развертывания купленных юнитов мгновенно или на следующем ходу
  в сценариях и кампании (2013/12/08 AndO3131)
- добавлен выбор для игры с / без основных блоков (2013/04/08 AndO3131)
- повышены лимиты для больших электронных файлов; размер листа значков ограничен h = 16384px
  SDL v1.x, а не LGD (2013/01/08 AGW)
- вы можете настроить электронный файл или другие текстовые файлы для использования любого сочетания WAV / OGG / MP3
  аудиофайлы, поддерживаемые микшером SDL (2013/01/08 AGW)
- вы можете настроить электронный файл или другие текстовые файлы для использования любого сочетания BMP / PNG / JPG
  растровые файлы, поддерживаемые SDL (2013/01/08 AGW)
- добавлены 4 новых типа движения, новый класс юнитов ДОТ и пример для
  новый тип железнодорожной местности; сделал разное. изменения единицы / местности (2013/01/08 AGW)
- базовая конфигурация cmake для основной программы (2013/11/11 AndO3131)
- 2 патча SourceForge (Galland)

Репозиторий SVN:
- улучшенные страницы руководства (2012/11/4 Маркус Кошаный)
- исправить сломанную дорожную плитку pg путем зеркалирования копии (2012/11/01 MS)
- удален общий часто задаваемый вопрос о non-free panzer (2012/10/29 MS)
- заменена / доработана несвободная графика / звуки (2012.10.29 Маркус Кошаный)
- применен патч hardened compile (2012/10/11 Маркус Кошаный)

1.2.3:
- не разрешать АИ размещать юниты на границе карты (2012/04/20 ТБ)
- назначать нации в подразделения поздней войны: FPO - Польша, FFR - Франция
  (2012/04/20 ТБ)
- исправлена ​​ошибка в окончательном просмотре сценария: правильно отключить опцию покупки
  поскольку в этом состоянии не выбран ни один игрок (17.04.2012 ТБ)
- добавлены финские идентификаторы для алгоритма "получить нацию по имени" в lgc-pg:
  он не используется PG, но некоторые дополнительные кампании (2012/01/27 MS)
- добавлены параметры lgc-pg для изменения имен игроков и наций (2012/01/27 MS)
- исправлена ​​ошибка преобразования тактических иконок (27.01.2012 MS)
- добавлены подсказки по сборке на Mac OS X (30.08.2011 ACV)
- генерировать имена файлов сценария из его заголовка для пользовательских кампаний
  (2011/08/17 мс)

1.2.2:
- применен ряд патчей сборки Gentoo (2011/08/16 MS)
- исправлена ​​ошибка в расчете дальности хода при одновременном использовании юнита и
  транспортер, не использовал топливо (2011.08.16 мс)
- исправлено переполнение буфера при генерации имени юнита (2011/08/16 MS)
- улучшена конвертация кампаний в lgc-pg (2011/08/09 MS)

1.2.1:
- диалог выбора видеорежима со всеми доступными размерами (13.06.2011 г.)

1.2:
- разрешить отключение / включение покупки и развертывания очереди в графическом интерфейсе (10.09.29 MS)
- отключить покупку при загрузке старых сохраненных игр (29.10.29 мс)
- улучшен бонус инициативы для русских танков (28.10.28 мс)
- использовать путь установки по умолчанию для lgc-pg (10.09.28 MS)
- используйте безопасный snprintf вместо sprintf в lgc-pg (10.09.28 MS)
- попробуйте открыть файлы PG с именем в нижнем и верхнем регистре (10.09.28 MS)
- добавлен заставка для lgc-pg (27.10.27 мс)
- добавлена ​​опция конфигурации для сообщений отладки AI (10.09.27 мс)
- добавлена ​​фаза покупки юнитов для AI (10.09.27 мс)
- получить престиж за захват флагов и нанесение урона юнитам (26.10.26 мс)
- добавлен диалог покупки (10.09 мс)
- lgc-pg строит данные престижа на основе исходных данных PG (10/09 MS)
- исправлено неправильное обновление центрального блока развертывания при загрузке сценария (10.09.22 MS)
- исправлена ​​компиляция на 64 битных системах Linux (09.10.20 Павол Руснак)
- исправлено переполнение буфера при загрузке кампании (10.09.20 Франк Шефер)

1.2бета-14:
- исправлено предупреждение о настройке (26.09.26 мс)
- добавлена ​​поддержка XDG (10.09.26 мс)
- исправлены некоторые ошибки в READMEs (01.09.09 BH)

1.2бета-13:
- исправлен неправильный цветовой ключ в растровых изображениях шрифтов (07.09.26 мс)
- добавлен оригинальный шрифт pg для брифингов кампании (07.09.24 Pamashoid)
- игнорировать возвращаемое значение в po / lgeneral Makefile (07.09.24 MS)
- исправление: использование транспортера вылетало из-за нехватки топлива (07.09.24 Triple
Entender)

1.2бета-12:
- исправлена ​​ошибка в gettext, приводившая к сбою make uninstall (29.06.29 LS)
- исправлена ​​ошибка, из-за которой end_turn блокировался, если NUMLOCK был активен (06.05.27 мс)
- скомпилировать под mingw. Спасибо Виктору Сергиенко. (04.06.10 LS)

1,2бета-11:
- начальное развертывание не разрешено для союзного игрока в MarketGarden (06.01.14 MS)
- исправлена ​​трансляция-кодирование с вызовом msgconv по мере необходимости (05.12.30 LS)
- обновлен немецкий перевод (12.05.29 LS)
- добавлена ​​опция --camp-start (12.05.29 LS)
- кампанию теперь можно загружать из командной строки (12.05.29 LS)
- добавлены полные описания для PG-кампании (12.05.28 LS)
- Приведены опасные зоны в соответствие с новыми правилами использования топлива (12.05.28 LS)

src/map.h
/***************************************************************************
                          map.h  -  description
                             -------------------
    begin                : Sat Jan 20 2001
    copyright            : (C) 2001 by Michael Speck
    email                : kulkanie@gmx.net
 ***************************************************************************/
/***************************************************************************
                     Modifications by LGD team 2012+.
 ***************************************************************************/
/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#ifndef __MAP_H
#define __MAP_H

#include "terrain.h"
#include "nation.h"
#include "player.h"
#include "unit.h"

#define MAX_LAYERS 5

enum { FAIR = 0, CLOUDS, RAIN, SNOW };

/*
====================================================================
Плитка карты
====================================================================
*/
typedef struct {
    char *name;             /* имя этого фрагмента карты */
    int terrain_id[MAX_LAYERS]; /* id свойств местности */
    int image_offset_x;     /* смещение изображения в prop->image */
    int image_offset_y;     /* смещение изображения в prop->image */
    int strat_image_offset; /* смещение в списке крошечных стратегических изображений местности */
    Nation *nation;         /* нация, владеющая этим флагом (NULL == нет нации) */
    Player *player;         /* вот */
    int obj;                /* военный объект? */
    int deploy_center;      /* развертывание разрешено? */
    Unit *g_unit;           /* указатель наземной / морской части */
    Unit *a_unit;           /* указатель воздушного блока */
    int layers[MAX_LAYERS]; /* слои этого фрагмента карты; слой 1 - дороги */
} Map_Tile;

/*
====================================================================
Для определения различных элементов карты (развертывание, обнаружение, блокировка ...)
маска карты используется, и это флаги для нее.
====================================================================
*/
enum {
    F_FOG =             ( 1L << 1 ),
    F_SPOT =            ( 1L << 2 ),
    F_IN_RANGE =        ( 1L << 3 ),
    F_MOUNT =           ( 1L << 4 ),
    F_SEA_EMBARK =      ( 1L << 5 ),
    F_AUX =             ( 1L << 6 ),
    F_INFL =            ( 1L << 7 ),
    F_INFL_AIR =        ( 1L << 8 ),
    F_VIS_INFL =        ( 1L << 9 ),
    F_VIS_INFL_AIR =    ( 1L << 10 ),
    F_BLOCKED =         ( 1L << 11 ),
    F_BACKUP =          ( 1L << 12 ),
    F_MERGE_UNIT =      ( 1L << 13),
    F_INVERSE_FOG =     ( 1L << 14 ), /* инверсия F_FOG */
    F_DEPLOY =          ( 1L << 15 ),
    F_CTRL_GRND =       ( 1L << 17 ),
    F_CTRL_AIR =        ( 1L << 18 ),
    F_CTRL_SEA =        ( 1L << 19 ),
    F_MOVE_COST =       ( 1L << 20 ),
    F_DANGER =          ( 1L << 21 ),
    F_SPLIT_UNIT =      ( 1L << 22 ),
    F_DISTANCE =        ( 1L << 23 )
};

/*
====================================================================
Map mask tile.
====================================================================
*/
typedef struct {
    int fog; /* если true, двигатель покрывает эту плитку туманом. если установлен ENGINE_MODIFY_FOG
                этот туман может меняться в зависимости от действия (дальность действия юнита, партнеров по слиянию
                и т.д */
    int spot; /* истина, если кто-либо из ваших юнитов наблюдает за этим тайлом карты; вы можете атаковать только юниты
    на плитке карты, которую вы заметили */
    /* используется для выбранного юнита */
    int in_range; /* это используется для поиска пути; это -1, если тайл не в диапазоне, иначе он установлен на
    оставшиеся движущиеся точки агрегата; влияние врага не учитывается */
    int distance; /* простое расстояние до текущего объекта; используется для маски опасности */
    int moveCost; /* общие затраты на достижение этой плитки */
    int blocked; /* юниты могут перемещаться туда вместе с союзным юнитом, но они не должны на этом останавливаться;
    поэтому разрешите переход к плитке, только если in_range и! заблокированы */
    int mount; /* истина, если юнит должен смонтировать, чтобы добраться до этой плитки */
    int sea_embark; /* возможен выход в море? */
    int infl; /* эта маска устанавливается в начале хода игрока; каждая плитка рядом с
    инфляция вражеского юнита увеличивается; если влияние равно 1, затраты на перемещение удваиваются; если влияние> = 2
    эта плитка непроходима (отряд останавливается на этой плитке); если юнит не видит плитку с infl> = 2 и
    пытается переместиться туда, он остановится на этой плитке; не зависит от движущихся точек юнита, проходящих через
    под влиянием тайла стоит все мов-поинты */
    int vis_infl; /* аналог infl, но только пятнистые единицы вносят вклад в эту маску; используется для установки
    маска in_range */
    int air_infl; /* аналог для летательных аппаратов */
    int vis_air_infl;
    int aux; /* используется для установки любого из верхних значений */
    int backup; /* используется для резервного копирования точечной маски для отмены перемещения юнита */
    Unit *merge_unit; /* если не NULL, это указатель на юнита, который вызвал map_get_merge_units ()
                         может сливаться с. вам нужно запомнить другой юнит, так как она не сохраняется здесь */
    Unit *split_unit; /* целевой отряд может передать силу */
    int  split_okay; /* юнит может передать новый подблок на этот тайл */
    int deploy; /* маска развертывания: 1: отряд может развернуть свои, 0 отряд не может их развернуть; настройка с помощью deploy.c */
    int danger; /* 1: отметьте эту плитку как опасную для входа */
    /* Маски ИИ */
    int ctrl_grnd; /* маска контролируемой зоны для игрока. собственные подразделения дают положительный бой
                      значение в движении + дальность атаки, в то время как очки врага вычитаются. финал
                      значение для каждой плитки относительно наивысшего абсолютного контрольного значения, таким образом
                      колеблется от -1000 до 1000 */
    int ctrl_air;
    int ctrl_sea; /* каждая рабочая область имеет свою маску управления */
} Mask_Tile;


/*
====================================================================
Загрузить карту.
====================================================================
*/
int map_load( char *fname );

/*
====================================================================
Удалить карту.
====================================================================
*/
void map_delete( );

/*
====================================================================
Получить плитку в x, y
====================================================================
*/
Map_Tile* map_tile( int x, int y );
Mask_Tile* map_mask_tile( int x, int y );

/*
====================================================================
Очистите переданные флаги маски карты.
====================================================================
*/
void map_clear_mask( int flags );

/*
====================================================================
Поменять местами юниты. Возвращает предыдущую единицу или 0, если нет.
====================================================================
*/
Unit *map_swap_unit( Unit *unit );

/*
====================================================================
Вставить, удалить указатель объекта с карты.
====================================================================
*/
void map_insert_unit( Unit *unit );
void map_remove_unit( Unit *unit );

/*
====================================================================
Получите соседние плитки по часовой стрелке с идентификатором от 0 до 5.
====================================================================
*/
Map_Tile* map_get_close_hex( int x, int y, int id );

/*
====================================================================
Добавить / установить засвет юнита во вспомогательную маску
====================================================================
*/
void map_add_unit_spot_mask( Unit *unit );
void map_get_unit_spot_mask( Unit *unit );

/*
====================================================================
Установите диапазон движения юнита на in_range / sea_embark / mount.
====================================================================
*/
void map_get_unit_move_mask( Unit *unit );
void map_clear_unit_move_mask();

/*
====================================================================
Воссоздает маску опасности для «юнита».
Для этого туман должен быть установлен на диапазон движения «юнит».
функция для правильной работы.
Стоимость движения маски должна быть установлена ​​как «юнит».
Возвращает 1, если установлена ​​хотя бы одна маска опасности для плитки, в противном случае - 0.
====================================================================
*/
int map_get_danger_mask( Unit *unit );

/*
====================================================================
Получите список путевых точек, по которым юнит движется к месту назначения.
Это включает проверку невидимого влияния вражеских юнитов (например,
Сюрприз-контакт).
====================================================================
*/
typedef struct {
    int x, y;
    int cost;
} Way_Point;



Way_Point* map_get_unit_way_points( Unit *unit, int x, int y, int *count, Unit **ambush_unit );

/*
====================================================================
Резервное копирование / восстановление точечной маски в / из резервной маски. Используется для отмены поворота.
====================================================================
*/
void map_backup_spot_mask();
void map_restore_spot_mask();

/*
====================================================================
Получить партнеров по слиянию юнита и установить маску слияния.
Максимум единиц MAP_MERGE_UNIT_LIMIT.
Все неиспользуемые записи в партнерах устанавливаются на 0.
====================================================================
*/
enum { MAP_MERGE_UNIT_LIMIT = 6 };
void map_get_merge_units( Unit *unit, Unit **partners, int *count );

/*
====================================================================
Проверьте, может ли юнит передать силу юниту (если не NULL) или создать
автономный модуль (если unit NULL) по координатам.
====================================================================
*/
int map_check_unit_split( Unit *unit, int str, int x, int y, Unit *dest );

/*
====================================================================
Получите разделенных партнеров юнита, предполагая, что юнит хочет дать силу 'str'
точки и установите маску «разделение». Максимум единиц MAP_SPLIT_UNIT_LIMIT.
Все неиспользованные записи в партнерах устанавливаются на 0. 'str' должно быть допустимым количеством,
здесь это не проверяется.
====================================================================
*/
enum { MAP_SPLIT_UNIT_LIMIT = 6 };
void map_get_split_units_and_hexes( Unit *unit, int str, Unit **partners, int *count );

/*
====================================================================
Получите список (vis_units) всех видимых единиц, проверив маску пятна.
====================================================================
*/
void map_get_vis_units( void );

/*
====================================================================
Нарисуйте на поверхности фрагмент карты местности. (затуманивается, если установлена ​​маска :: туман)
====================================================================
*/
void map_draw_terrain( SDL_Surface *surf, int map_x, int map_y, int x, int y );
/*
====================================================================
Нарисуйте блоки плитки. Если установлена ​​маска :: fog, юниты не отрисовываются.
Если 'ground' имеет значение True, наземный блок отображается как основной.
и воздушный блок нарисован маленьким (и наоборот).
Если установлено 'select', добавляется рамка выбора.
====================================================================
*/
void map_draw_units( SDL_Surface *surf, int map_x, int map_y, int x, int y, int ground, int select );
/*
====================================================================
Нарисуйте плитку опасности. Ожидает, что "surf" будет содержать полностью нарисованную плитку в
данная позиция, которая будет окрашена наложением опасности
поверхность местности.
====================================================================
*/
void map_apply_danger_to_tile( SDL_Surface *surf, int map_x, int map_y, int x, int y );
/*
====================================================================
Нарисуйте местность и юниты.
====================================================================
*/
void map_draw_tile( SDL_Surface *surf, int map_x, int map_y, int x, int y, int ground, int select );

/*
====================================================================
Установить / обновить маску пятна текущим игроком движка.
Обновление добавляет плитки, видимые юнитом.
====================================================================
*/
void map_set_spot_mask();
void map_update_spot_mask( Unit *unit, int *enemy_spotted );

/*
====================================================================
Установите маска :: туман (который является фактическим туманом двигателя) либо
точечная маска, маска in_range (закрывает sea_embark), маска слияния,
развернуть маску.
====================================================================
*/
void map_set_fog( int type );

/*
====================================================================
Установите туман на маску пятна игрока, используя маска :: aux (не маска :: место)
====================================================================
*/
void map_set_fog_by_player( Player *player );

/*
====================================================================
Убедитесь, что этот фрагмент карты виден движку (не покрыт
by mask :: fog или mask :: spot, так как модификация разрешена и может быть
туман другого игрока (например, один человек против процессора))
====================================================================
*/
#define MAP_CHECK_VIS( mapx, mapy ) ( ( !modify_fog && !mask[mapx][mapy].fog ) || ( modify_fog && mask[mapx][mapy].spot ) )

/*
====================================================================
Измените различные маски влияния.
====================================================================
*/
void map_add_unit_infl( Unit *unit );
void map_remove_unit_infl( Unit *unit );
void map_remove_vis_unit_infl( Unit *unit );
void map_set_infl_mask();
void map_set_vis_infl_mask();

/*
====================================================================
Проверить, может ли установка производить посадку в воздух / море / высадку в точке x, y.
Если 'init'! = 0, используются упрощенные правила для развертывания
====================================================================
*/
int map_check_unit_embark( Unit *unit, int x, int y, int type, int init );
int map_check_unit_debark( Unit *unit, int x, int y, int type, int init );

/*
====================================================================
Посадить / высадить юнит и вернуться, если был обнаружен враг.
Если «вражеское пятно» равно 0, не пересчитывать маску пятна.
Если координаты объекта или x и y выходят за границы, соответствующие
плиткой не манипулируют.
====================================================================
*/
void map_embark_unit( Unit *unit, int x, int y, int type, int *enemy_spotted );
void map_debark_unit( Unit *unit, int x, int y, int type, int *enemy_spotted );

/*
====================================================================
Установите маску развертывания для этого устройства. Если 'init', используйте начальное развертывание
маска (или по умолчанию). Если нет, установите действующие центры развертывания. В
Во втором прогоне удалите любую плитку, заблокированную собственным юнитом, если установлено значение «unit».
====================================================================
*/
void map_get_deploy_mask( Player *player, Unit *unit, int init );

/*
====================================================================
Отметьте, что это поле является полем развертывания для данного игрока.
====================================================================
*/
void map_set_deploy_field( int mx, int my, int player );

/*
====================================================================
Проверьте, является ли это поле полем развертывания для данного игрока.
====================================================================
*/
int map_is_deploy_field( int mx, int my, int player );

/*
====================================================================
Проверьте, можно ли развернуть модуль на mx, my или вернуть неразвертываемый модуль
там. Если установлен air_mode, сначала проверяется воздушный блок.
«игрок» - это индекс игрока.
====================================================================
*/
int map_check_deploy( Unit *unit, int mx, int my, int player, int init, int air_mode );
Unit* map_get_undeploy_unit( int x, int y, int air_region );

/*
====================================================================
Проверьте уровень поставки плитки (mx, my) в контексте «единицы».
(шестиугольные плитки с SUPPLY_GROUND имеют 100% запас)
====================================================================
*/
int map_get_unit_supply_level( int mx, int my, Unit *unit );
/*
====================================================================
Проверьте, является ли этот фрагмент карты точкой снабжения данного отряда.
====================================================================
*/
int map_is_allied_depot( Map_Tile *tile, Unit *unit );
/*
====================================================================
Проверяет, поставляется ли этот гексагон (mx, my) складом в
контекст «юнита».
====================================================================
*/
int map_supplied_by_depot( int mx, int my, Unit *unit );

/*
====================================================================
Получить зону высадки для юнита (все близкие гексы, которые свободны).
====================================================================
*/
void map_get_dropzone_mask( Unit *unit );

#endif
